Accept more numpy scalars as doubles

t-vi opened the pull request:
Allows mulitplication of e.g. numpy.float32 with tensors.

This came up with #9468

If you want this and after the other patch is done, I'll add tests (but that would be conflicting, so I prefer to wait).

t-vi committed:
Accept more numpy scalars as doubles

Allows mulitplication of e.g. numpy.float32 with tensors.
This came up with #9468

soumith commented:
@pytorchbot retest this please

gchanan commented:
looks good!  Maybe add a couple of tests?

t-vi commented:
I would want to test this by adding numpy float32 scalars to the test introduced in #9651, so my original plan was to revise this PR after that to avoid the merge conflict.

t-vi committed:
add test with mul

SsnL commented on test/test_torch.py:
```diff
@@ -7608,6 +7608,15 @@ def test_numpy_array_interface(self):
+        for np_dtype in [np.float32, np.float64, np.int32, np.int64]:
+            np_sc = np_dtype(2.0)
+            t = torch.ones(2, requires_grad=True)
+            r1 = t * np_sc
```
Should `r1` all be `double`s (`t` is `double` because of default dtype is set as double in test)? Can you add an assert on the dtype?

t-vi committed:
Check for torch dtypes float and double and also assert result dtype

From SsnL's review. Thank you!

weiyangfb commented:
@t-vi needs a rebase now :( Is this ready for merging now? @SsnL @gchanan

yf225 commented:
@t-vi We probably need to rebase this PR onto current master

t-vi commented:
Will do next week.

t-vi committed:
Merge branch 'master' into accept_more_scalars

weiyangfb commented:
is this ready to go? @gchanan @SsnL

SsnL commented:
LGTM. @gchanan any other thoughts?

gchanan commented on test/test_torch.py:
```diff
@@ -8192,14 +8192,18 @@ def test_numpy_array_interface(self):
-        r2 = t * np_sc
-        self.assertIsInstance(r2, torch.Tensor)
-        self.assertTrue(r2.requires_grad)
+        for np_dtype in [np.float32, np.float64, np.int32, np.int64]:
```
does it work to test int16, uint8 here?

gchanan commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -85,9 +86,9 @@ inline int64_t THPUtils_unpackIndex(PyObject* obj) {
 inline bool THPUtils_checkDouble(PyObject* obj) {
 #if PY_MAJOR_VERSION == 2
-  return PyFloat_Check(obj) || PyLong_Check(obj) || PyInt_Check(obj);
+  return PyFloat_Check(obj) || PyLong_Check(obj) || PyInt_Check(obj) || torch::utils::is_numpy_scalar(obj);
```
don't you want to guard the call to is_numpy_scalar here?  Because this will throw an exception if we didn't compile with numpy and the object isn't a PyFloat, PyLong, PyInt?

gchanan commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -85,9 +86,9 @@ inline int64_t THPUtils_unpackIndex(PyObject* obj) {
+  return PyFloat_Check(obj) || PyLong_Check(obj) || PyInt_Check(obj) || torch::utils::is_numpy_scalar(obj);
 #else
-  return PyFloat_Check(obj) || PyLong_Check(obj);
+  return PyFloat_Check(obj) || PyLong_Check(obj) || torch::utils::is_numpy_scalar(obj);
```
and here.

t-vi replied:
Thank you for these, I'll update the PR.

t-vi committed:
Guard numpy conversion with USE_NUMPY ifdefs

And expand range of tested numpy types.
Thank you, gchanan for the suggestions!

t-vi commented:
Is the `TestAdam.test_sparse_adam_output_grad` caused by the PR? At first sight, I'm not entirely sure I see the relation to the changes.

weiyangfb committed:
Merge branch 'master' into accept_more_scalars

weiyangfb commented:
Looks like comments are addressed, should we merge this? @gchanan

gchanan reviewed (approved):
lgtm.
